File: Source/code/forms/formToggleLockStatus.pas
// ===================================================================
//
// (c) Paul Alan Freshney 2012-2022
// www.freshney.org :: paul@freshney.org :: maximumoctopus.com
//
// https://github.com/MaximumOctopus/LEDMatrixStudio
//
// Please do not modifiy this comment section

//
// ===================================================================


unit formToggleLockStatus;


interface


uses
  Winapi.Windows, Winapi.Messages, System.SysUtils, System.Variants, System.Classes, Vcl.Graphics,
  Vcl.Controls, Vcl.Forms, Vcl.Dialogs, Vcl.StdCtrls, Vcl.Buttons, Vcl.ExtCtrls,

  languagehandler;


type
  TfrmToggleLockStatus = class(TForm)
    Bevel1: TBevel;
    bOK: TBitBtn;
    bCancel: TBitBtn;
    Image1: TImage;
    GroupBox1: TGroupBox;
    Label1: TLabel;
    Label2: TLabel;
    eStartFrame: TEdit;
    eEndFrame: TEdit;
    cbLockStatus: TCheckBox;
    procedure eStartFrameChange(Sender: TObject);
    procedure FormCreate(Sender: TObject);
  private
    FFrameCount : integer;

    procedure SetGUILanguageText;

    function ValidateInputs: boolean;
  end;


  TToggleLockFrameRangeObject = record
    Process              : boolean;
    LockStatus           : boolean;
    StartFrame, EndFrame : integer;
  end;


var
  frmToggleLockStatus: TfrmToggleLockStatus;


function DoToggleLockFrameRange(aFrameCount : integer): TToggleLockFrameRangeObject;


implementation


{$R *.dfm}


function DoToggleLockFrameRange(aFrameCount : integer): TToggleLockFrameRangeObject;
begin
  with TfrmToggleLockStatus.Create(Application) do
    try
      FFrameCount := aFrameCount;

      Result.Process := False;

      ShowModal;

      if ModalResult = mrOK then begin
        Result.Process    := True;
        Result.LockStatus := cbLockStatus.Checked;
        Result.StartFrame := StrToInt(eStartFrame.Text);
        Result.EndFrame   := StrToInt(eEndFrame.Text);
      end;

    finally
      Free;
    end;
end;


procedure TfrmToggleLockStatus.FormCreate(Sender: TObject);
begin
  SetGUILanguageText;
end;


procedure TfrmToggleLockStatus.SetGUILanguageText;
begin
  Caption := GLanguageHandler.Text[kToggleFrameLockStatus];
  Label1.Caption := GLanguageHandler.Text[kStart];
  Label2.Caption := GLanguageHandler.Text[kEnd];
  cbLockStatus.Caption := GLanguageHandler.Text[kLock];

  bOK.Caption := GLanguageHandler.Text[kOK];
  bCancel.Caption := GLanguageHandler.Text[kCancel];
end;


procedure TfrmToggleLockStatus.eStartFrameChange(Sender: TObject);
begin
  bOk.Enabled := ValidateInputs;
end;


function TfrmToggleLockStatus.ValidateInputs: boolean;
var
  sf, ef : integer;

begin
  sf := StrToIntDef(eStartFrame.Text, -1);
  ef := StrToIntDef(eEndFrame.Text, -1);

  Result := (sf <> -1) and (ef <> -1) and (sf <= ef);
end;


end.

File: Source/Forms/FormToggleLockStatus.cpp
// ===================================================================
//
//   (c) Paul Alan Freshney 2012-2024
//   www.freshney.org :: paul@freshney.org :: maximumoctopus.com
//
//   https://github.com/MaximumOctopus/LEDMatrixStudio
//
//   https://maximumoctopus.hashnode.dev/
//
//   C++ Rewrite October 11th 2023
//
// ===================================================================

#include <vcl.h>
#pragma hdrstop

#include "FormToggleLockStatus.h"
#include "LanguageConstants.h"
#include "LanguageHandler.h"

#pragma package(smart_init)
#pragma resource "*.dfm"
TfrmToggleLockStatus *frmToggleLockStatus;

extern LanguageHandler *GLanguageHandler;

//---------------------------------------------------------------------------
//---------------------------------------------------------------------------

ToggleLockFrameRange OpenToggleLockStatus()
{
	TfrmToggleLockStatus *frmToggleLockStatus = new TfrmToggleLockStatus(Application);

	ToggleLockFrameRange tlfr;

	if (frmToggleLockStatus->ShowModal() == mrOk)
	{
		tlfr.Process = true;
		tlfr.LockStatus = frmToggleLockStatus->cbLockStatus->Checked;
		tlfr.StartFrame = frmToggleLockStatus->eStartFrame->Text.ToIntDef(1);
		tlfr.EndFrame   = frmToggleLockStatus->eEndFrame->Text.ToIntDef(1);
	}

	delete frmToggleLockStatus;

	return tlfr;
}

//---------------------------------------------------------------------------
//---------------------------------------------------------------------------


__fastcall TfrmToggleLockStatus::TfrmToggleLockStatus(TComponent* Owner)
	: TForm(Owner)
{
	SetGuiLanguageText();
}


void __fastcall TfrmToggleLockStatus::eStartFrameChange(TObject *Sender)
{
	bOk->Enabled = ValidateInputs;
}


bool TfrmToggleLockStatus::ValidateInputs()
{
	int sf = eStartFrame->Text.ToIntDef(-1);
	int ef = eEndFrame->Text.ToIntDef(-1);

	return (sf != -1 && ef != -1 && sf <= ef);
}


void TfrmToggleLockStatus::SetGuiLanguageText()
{
	Caption = GLanguageHandler->Text[kToggleFrameLockStatus].c_str();
	Label1->Caption = GLanguageHandler->Text[kStart].c_str();
	Label2->Caption = GLanguageHandler->Text[kEnd].c_str();
	cbLockStatus->Caption = GLanguageHandler->Text[kLock].c_str();

	bOk->Caption = GLanguageHandler->Text[kOK].c_str();
	bCancel->Caption = GLanguageHandler->Text[kCancel].c_str();
}
